drivers/sensor/voltage_divider/voltage.c

```diff
@@ -59,6 +59,12 @@ static int get(const struct device *dev, enum sensor_channel chan, struct sensor
 		return -ENOTSUP;
 	}
 
+	if (config->voltage.port.channel_cfg.differential) {
+		raw_val = (int16_t)data->raw;
```

Comment:
At the moment this will break compatibility with nRF ADCs and possibly other ADCs that can return negative results in single-ended mode. Can we add something like:
```suggestion
	if (config->voltage.port.channel_cfg.differential ||
	    config->signed_single_ended) {
		raw_val = (int16_t)data->raw;
```

Reply:
This seems to be expanding into a bigger issue than the voltage divider sensor.

In what circumstances does a single-ended nRF ADC channel give a negative value? I took a quick glance at some online documentation but failed to find it. Also I didn't find any mention of it in the adc_nrf*.c code.

Meanwhile, include/zephyr/drivers/adc.h actually defines single-ended channels as inherently unsigned:

```	/**
	 * ADC resolution.
	 * For single-ended channels the sample values are from range:
	 *   0 .. 2^resolution - 1,
	 * for differential ones:
	 *   - 2^(resolution-1) .. 2^(resolution-1) - 1.
	 */
```

If we want to change this definition, then it sounds like we need a bigger refactor of Zephyr ADC code and perhaps a new Kconfig option. But do we really want that, or is this a nRF-specific quirk that should be handled in nRF-specific driver code?

Reply:
The nRF ADC works over the full scale input range (0V to VDD) so any non linearity and/or offsets in the hardware when measuring near 0V can cause the sample to be negative.

At the moment we don't have support for greater than 16-bit ADCs either. In the short term to fix this issue for <=16-bit ADCs we could do the following:

```suggestion
	if (config->voltage.port.channel_cfg.differential || (config->voltage.port.resolution < 16)) {
		raw_val = (int16_t)data->raw;
```


From **nRF52832_PS_v1.9.pdf**
<img width="874" alt="image" src="https://github.com/zephyrproject-rtos/zephyr/assets/131832/c2570830-c0ee-485e-8fef-4591ff203dfe">

Reply:
Aha, now I understand:

> _"Internally, the ADC is always a differential analog-to-digital converter, but by default it is configured with single-ended input in the MODE field of the CH[n].CONFIG register. In single-ended mode, the negative input will be shorted to ground internally."_
> -- https://infocenter.nordicsemi.com/topic/com.nordic.infocenter.nrf52832.ps.v1.1/saadc.html

This explains why it can return negative values in single-ended mode: It's still differential.

Isn't it then up to the nRF ADC driver to make sure that the the device behaves as a single-ended ADC when configured as such, i.e don't return differential values? Alternatively, to not allow configuring it as a single-ended ADC, when it de-facto isn't.

The problem is that this "surprise sign bit" for single-ended readings breaks assumptions in more places than this sensor. A quick grep for "uint16_t raw" reveals other sensors that assume the adc value is unsigned and will break if a sign bit is injected. I suspect there are more places still that just require a little more effort to find. And we have no place in the API to specify that a purportedly single-ended ADC is actually differential.

That elevates this issue to an archictecture/API question rather than just a simple sensor bug fix. I'd like to ping in @MaureenHelm (Sensors maintainer) and @anangl (ADC & nRF maintainer) to get their insights.

Reply:
> Isn't it then up to the nRF ADC driver to make sure that the the device behaves as a single-ended ADC when configured as such, i.e don't return differential values? Alternatively, to not allow configuring it as a single-ended ADC, when it de-facto isn't.

Yes, this should be corrected in the driver. It should not return negative values in the single-ended mode.

Reply:
> > _"Internally, the ADC is always a differential analog-to-digital converter, but by default it is configured with single-ended input in the MODE field of the CH[n].CONFIG register. In single-ended mode, the negative input will be shorted to ground internally."_
> > -- https://infocenter.nordicsemi.com/topic/com.nordic.infocenter.nrf52832.ps.v1.1/saadc.html
> 
> This explains why it can return negative values in single-ended mode: It's still differential.

👍

Reply:
I'm not sure how quickly https://github.com/zephyrproject-rtos/zephyr/issues/71119 will be resolved so the following could be used to provide 16-bit single-ended support without breaking nRF support for the driver:

```suggestion
	if (config->voltage.port.channel_cfg.differential) {
		raw_val = (int16_t)data->raw;
	} else if (config->voltage.port.resolution < 16) {
		/* Can be removed when https://github.com/zephyrproject-rtos/zephyr/issues/71119 is resolved */
		raw_val = (int16_t)data->raw;
```

Reply:
Right. I'm not sure what the normal way-of-working is in Zephyr. Do we merge temporary workarounds around pending issues or do we try to focus on getting the root cause fixed? Do we have any guesstimate on the timeframe for #71119?

As mentioned, this issue is affecting more places than just this voltage-divider sensor. It would feel a little awkward adding a workaround here but leaving other places broken/impacted. On the other hand finding all affected places may be a non-trivial task and a big/noisy patch followed (hopefully) by a revert shortly.

Advice welcome.

Reply:
Based on the code review a subsystem maintainer makes the decision when to merge. As a user of this driver I'm happy to wait for the resolution of #71119. Also happy for my suggestion above to be used. I've seen similar comments in Zephyr code before so I believe the suggestion code would be ok. I would definitely prefer my usage not to be broken in the next release in the case #71119 was not resolved in time.